### ISSUE TITLE
Statically-Typed Keys

### ISSUE THREAD

**Issue by paddycarver** [json-api/json-api#8]
The ambiguity around the type of the keys (namely, that they can be either single objects or an array of objects) presents a problem for most statically-typed languages. For example, in Android, I always build deserialization right into the network request layer. As part of this, I build a generic "Response" class that the JSON is deserialized into. I can then interact with that Response class to obtain the data in a statically typed way, which ensures type-safety within the application itself.

While this project is geared towards Javascript applications, I can't build an API against a schema I know won't play well with mobile or other clients I may write. So I think this is an issue worth resolving.

I have two proposals for a solution:

We can use plurality to separate between types; e.g., a single post could be returned as:

```
{
  "post": {
    "id": 1
  }
}
```

and multiple posts could be returned as:

```
{
  "posts": [
    { "id": 1 }, { "id": 2 }
  ]
}
```

The problem with this is that clients now have two places to look for the content they're expecting (if I GET /posts and it only returns one post, should that be in "post" or "posts"?) and the vagaries of pluralisation are now an API concern.

The alternative proposal (which is what I use in my APIs) is to never return a bare object. Whether one post or one hundred posts are returned, it is always under the "posts" key, and it is always in an array:

```
{
  "posts": [
    { "id": 1 }
  ]
}
```

This allows for type-safety by the simple addition of two characters to the response. It always locates the content under the same key, regardless of any other conditions. And it adds no new overhead or meaning to keys. The only drawback is that there is a slight cognitive dissonance around asking for one resource ( GET /posts/1) and receiving a list in response. I would argue, though, that semantically this response is correct: the user is not requesting the post object itself, the user is requesting a response object containing the post object--which is why it's semantically permissible to include the metadata and errors in the response, as well.

**Comment by steveklabnik** [comment]
I generally prefer the 'always return an array' version, yeah. I'm not sure if @wycats has a preference here.

**Comment by wycats** [comment]
"Always return an Array" seems good.

There is one small issue: It is possible for a request for a single post to come back with multiple posts (if they're related or preemptively provides).

There are cases where it's important to link up a particular document in a response to the request, most notably after a create when the response contains an ID to use for the newly created record. There is also a case where a document was `GET`ed at `/people/me`, but the response document contains a canonical ID that should be used instead.

I'm not entirely sure how to link up the "main" document back to the request, since in both of those cases the document's URL was not known when the original request was made. The current Ember Data approach handles this by single/plural, but I specifically tried to remove single/plural in JSON API to eliminate the need for coordinated inflectors (or hardcoded inflections) on both sides.

Thoughts?

**Comment by dgeb** [comment]
I like a lot about the "always return an array" approach. It shares the advantages I mentioned in #16, while also being semantically correct (even if it's still a bit unconventional).

@wycats - I think the issues that you've raised are also present in the current spec in which a plural key can be used to represent a single object or array of objects. Imagine creating a `person` and requesting that person's ancestry be included in the response. You'd again have an array of `people`, with the same problem of trying to discern the newly created one from the ancestors.

I think that what we're really discussing is a strategy for specifying separate statuses for components of compound documents. This would allow clients to identify which records are newly created, updated, or simply retrieved. I think this should tie into the error discussion for compound documents (#7). A well thought out solution will pave the way for conventions for batch operations.

**Comment by paddycarver** [comment]
I'm not a huge fan of `/people/me` URLs (they set up special URL rules for no benefit that I can see), but that's a discussion for another issue.

It seems the issue at hand here is how to determine the "main" resource and "auxiliary" resources when those resources share a type and the client does not know the ID of the main resource ahead of time (as it would for, for example, GET /posts/2).

Again, I have multiple proposals.

First, we could insert this information in the meta information returned with the request, with the pagination. That would be a clear and simple, if not particularly elegant, solution to the issue at hand.

Second, as @dgeb suggested, we could return meta information with each object about whether it was created or not. I'm not a huge fan of this solution, as it is easy to bloat that information, and the "status of the resource" is ostensibly what the status code is supposed to be representing in the first place. Also, in situations in which a master-slave relationship is created, both resources are created at once, which again confuses which resource is the primary one.

Third, and my preferred solution, is to assign meaning to the orders of the arrays. Quite simply, the first item in any array is always the primary object, should that resource type need a primary type for this response. I think this also matches up to how the data is created on the backend, and therefore would generally be how information is returned without any convention whatsoever: the resource is created and added to the response, the resource's auxiliary resources are created or queried, and they are added to the response. The drawback to this is that it fills the single slot we have to assign meaning to array orders, it's extra work when a request requires a sorted response, and it requires us to define semantics around when a response requires a primary object vs. a response is simply a list of objects with no primacy.

Did I miss an option?

**Comment by hjdivad** [comment]
@paddyforan One possibility is to combine your third and first options.  When present, well known keys in `meta` are used to determine primary documents.  When not present, ambiguity is resolved with order.

This lets people take a very simple approach for common cases but still have a relatively easy fallback when ordering the primary document is painful.

The downside is it adds complexity to the client, although I think the complexity is minimal.

**Comment by paddycarver** [comment]
@hjdivad the question then becomes how do you determine between a response that simply has no priority (e.g., "here are a bunch of comments, obviously none of them is the master or primary comment") instead of a response that you should imply the priority (e.g., "here are a bunch of comments, and the first one is the one you requested, but the others are all related").

The solution for most examples I keep arriving at is that that information should be part of the resource, not the API structure. E.g., a comment and child comments. I think in most my use cases, I am simply syncing a client-side database or state with the server-side state, then using that client-side state to navigate relationships, etc. So I can't think of any use cases that run afoul of the solution you proposed (or even just the third option I proposed).

**Comment by hjdivad** [comment]
@paddyforan at least in some cases, that there is a primary document is implied.  Consider the example given by @wycats of getting a document back after making a `POST`.  There is exactly one document that contains the id for the newly created record, a conforming server must do _something_ to indicate which document that is, either by providing only one document or by obeying some convention (e.g. primary document first).

His other example, `GET` `/people/me` also has a primary document.

At present, I can't think of a use case for a request whose response may or may not have a primary document.  I'm tempted to have the spec punt on this issue.  For example, the spec could read something like the following:

> If there is a primary document at the requested URL, and its ordinal position is not specified by `meta.primary`, the primary document must be the first document listed

The spec could then, in some circumstances, indicate when a  primary document is required (e.g. after `POST` `/resource`).  Other cases, such as `GET` `/people/me` require unspec-d coordination between client and server.

**Comment by paddycarver** [comment]
@hjdivad Right, I understand the theoretical problem posed, but I'm having trouble coming up with use cases that return multiple objects of a single type when the primary object is not known. The only times I can think of when the primary object is not known is during creation and using URL hacks like `/people/me`. In both cases, I can't think of a use case that would require me to return multiple objects of the same type. Which is not to assert that there are none. I'm hoping someone will give me one so that I can wrap my brain around it and consider what my expectations, as a consumer, would be.

**Comment by hjdivad** [comment]
@paddyforan One example is `POST /resource` automatically creates a child resource, which is returned in the original request.

**Comment by paddycarver** [comment]
@hjdivad Right. But what is "resource"? What tangible example is there of creating an object explicitly that implicitly creates an object of the same type?

The best I have is if I `POST /posts` and the application (for whatever reason) has a post.links field that keeps track of which posts any given post links to.

Personally, my expectation (as a client) would be that I would receive only the post I just `POST`ed (perhaps I should have chosen a different resource type) instead of the post and every post it links to. Though that may be a discussion more closely tied to #19. In that case, I think either of our propositions addresses the problem reasonably, and the client already knows it expects a primary object.

**TL;DR**: can anyone think of a situation in which the client doesn't already know it wants a primary object and in which multiple objects of the same type can be returned?

**Comment by hjdivad** [comment]
@paddyforan consider `POST /forum/1/message?crosspost=2,3,4`.  In this case, as a client, I may expect that this actually creates four posts with unique ids and I would want all of them back.  There is still a primary post: the one to forum 1.

**Comment by paddycarver** [comment]
Maybe I'm crazy, but I'd say in that example, the client should be making 4 requests. But in that case, yes, I think our proposed solutions hold up, as the client knows it is going to be receiving multiple objects in the response and that one of them needs to be designated as the primary.

**Comment by steveklabnik** [comment]
The way we currently add types in AMS looks like this, for reference. https://github.com/rails-api/active_model_serializers/commit/72652ce37e4a6fe538c5eb9afe3af9e6591121d8

**Comment by paddycarver** [comment]
@steveklabnik that just looks like the server-side portion to me and doesn't actually show the API representation. Did my non-Ruby background betray me, or were you just commenting on how to specify the type of the variable in Ruby?

**Comment by steveklabnik** [comment]
Yeah, just adding context for the server-side part. Servers and clients are part of this discussion. ;)

Since this spec is built from running code, I figure the context of 'here's how this is implemented now' can help. It's more general, not making a point about what you said specifically.

**Comment by dgeb** [comment]
Wow, a guy goes to mow the lawn and misses all of this ;)

@paddyforan - I didn't necessarily say _how_ meta information should be included about each resource. I agree that the status code should apply to the primary document unless we're designing a batch-oriented API spec (but it couldn't hurt to do think about those concepts at the same time).

The clearest method that I can think of to separate the primary document(s) from ancillary documents is to allow for a top level `rels` section to contain all the ancillary documents, which are all "related" to the primary document after all. This parallels the new (to AMS) `rels` section within each resource:

```
{
  "posts": [{
    "id": "1",
    "title": "Rails is Omakase",
    "rels": {
      "author": 9
    }
  }],
  "rels": {
    "people": [{
      "id": "9",
      "name": "@d2h"
    }]
  }
}
```

There are two downsides to this approach:
- it's possible (but extremely unlikely) that the same record may appear at the root level and in `rels`
- the top-level `rels` section is currently used for specifying a URL template and might need to be rethought to avoid conflicts

However, it would eliminate confusion over what record(s) are primary. Therefore, it would also be obvious what the status codes apply to.

**Comment by lukfugl** [comment]
I support @dgeb's proposition that related documents of the same type as the primary document be separated from the primary document. whether `rels` is the right place for that would still be an open question (it sounds good to me at first blush, but I haven't thought deeply about it yet).

**Comment by indirect** [comment]
Related to the sorting ticket I just opened at #22, it's possible that the order of the returned array is significant -- in that case, simply making it the first returned object no longer suffices to denote that it is the canonical or primary response object. In the last real-life API I worked on that had this exact problem, we wound up returning a separate top-level key that included the name and ID of the canonical object, so that the client didn't have to do any inflecting or guessing.

**Comment by mahmoudimus** [comment]
A client that's interpreting the response from a server like this:

``` js
{
  "posts": {
    "id": "1",
    "title": "Rails is Omakase",
    "rels": {
      "author": "http://example.com/people/1",
      "comments": "http://example.com/comments/5,12,17,20"
    }
  }
}
```

has to also interpret a response like this:

``` js
{
  "posts": {
    "id": "1",
    "title": "Rails is Omakase",
    "rels": {
      "comments": "http://example.com/posts/1/comments"
    }
  }
}
```

Notice that `posts["rels"]["comments"]` for the first example and `posts["rels"]["comments"]` are just blobs of strings, but there is nothing describing what's residing at those links.

In the first example, an array of exactly 4 times are coming back. In the second example, it could be an infinite page with many potentially different types of comments intermingled (as mentioned here #28).

How would a client know how to interpret those items? Is the URI referencing a page? Is it referencing a particular resource?

At @balanced, we solved this problem like [this](https://docs2.balancedpayments.com/current/api?language=bash#id24):

``` js
{
    "_type": "page", 
    "_uris": {
        "first_uri": {
            "_type": "page", 
            "key": "first"
        }, 
        "last_uri": {
            "_type": "page", 
            "key": "last"
        }, 
        "next_uri": {
            "_type": "page", 
            "key": "next"
        }, 
        "previous_uri": {
            "_type": "page", 
            "key": "previous"
        }
    }, 
    "items": [
        {
            "_type": "bank_account", 
            "_uris": {
                "credits_uri": {
                    "_type": "page", 
                    "key": "credits"
                }, 
                "verifications_uri": {
                    "_type": "page", 
                    "key": "verifications"
                },
            "routing_number": "121042882", 
            "type": "checking", 
            "uri": "/v1/bank_accounts/BA6jbpWWJDdUTocOWyhvrCGP", 
            "credits_uri": "/v1/bank_accounts/BA6jbpWWJDdUTocOWyhvrCGP/credits", 
            "verifications_uri": "/v1/bank_accounts/BA6jbpWWJDdUTocOWyhvrCGP/verifications"
            },
        } /* and so on an so forth */
   ]
  "last_uri": "/v1/bank_accounts?limit=2&offset=0", 
    "limit": 2, 
    "next_uri": null, 
    "offset": 0, 
    "previous_uri": null, 
    "total": 2, 
    "uri": "/v1/bank_accounts?limit=2&offset=0"
} 
```

This tells the client, when it sees a `credits_uri`, it will construct an `Page` class under the key `credits`. That way, when writing clients, you can do:

``` ruby
bank_account = BankAccount.find(:bank_account_uri)
bank_account.credits   # generates a Page object that will allow lazy infinite iteration of the related credits resource
```

It's very unambiguous and solves the problem. How would I solve this same problem using JSON API spec without having to resort to using client side code to detect what "type" I am going to instantiate for that particular URI?

**Comment by SphtKr** [comment]
Looks like the plan one way or another is to always use the plural version of the key (returning a single-item array or just a document), so on a side note this should probably be the case for URL template placeholders also? The current spec gives the example: 

```
"rels": {
    "posts.comments": "http://example.com/posts/{post.id}/comments"
}
```

but will it be `{posts.id}` instead?

**Comment by paddycarver** [comment]
The "rels" top-level object leaves me with a funny feeling--it's almost like an API smell, for some reason, just doesn't feel semantic--but it's the most elegant solution to the problem I've seen thus far. The biggest concern I have is that it, once again, means you have to look in multiple places to get the resources you're looking for from the request.

I could be reading @mahmoudimus's comment wrong, but that feels like a pagination concern to me. Which is a bit of a different conversation?

The thing I'm really asking for here is that in the scope of every object (and sub-object), the same key will always be associated with the same type. So polymorphism breaks that pretty badly, and is something I'd argue should be avoided.

**Comment by steveklabnik** [comment]
> The thing I'm really asking for here is that in the scope of every object (and sub-object), the same key will always be associated with the same type. So polymorphism breaks that pretty badly, and is something I'd argue should be avoided.

As I said in the other thread, I agree.

I'm updating this to always wrap things in arrays, and @wycats and I have one or two edge cases with this still, but we can fix those at a later time.
